lib/systems: add emulatorAvailable

Artturin opened the pull request:
```
nix-repl> pkgsCross.arm-embedded.stdenv.hostPlatform.emulatorAvailable pkgsCross.arm-embedded.buildPackages
false

nix-repl> pkgsCross.aarch64-multiplatform.stdenv.hostPlatform.emulatorAvailable pkgsCross.aarch64-multiplatform.buildPackages
true
```

will be useful for stuff like handling https://github.com/NixOS/nixpkgs/issues/187109

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sternenseemann commented on lib/systems/default.nix:
```diff
@@ -189,6 +189,9 @@ rec {
         then "${pkgs.mmixware}/bin/mmix"
         else throw "Don't know how to run ${final.config} executables.";
 
+        emulatorAvailable = pkgs:
```
I'd rather replace the `throw` with `null` – that will also fail any string interpolation (which is required for emulator always anyways).

`tryEval` is arguably a bit of a code smell and it will also not always work, e.g. often `targetPackages` will be basically empty which we also can't catch (but could in the emulator function).

Artturin replied:
how about now?

i want to keep the throw to have a clear message of why it fails instead of just `error: cannot coerce null to a string`

Artturin committed:
lib/systems: add emulatorAvailable

```
nix-repl> pkgsCross.arm-embedded.stdenv.hostPlatform.emulatorAvailable pkgsCross.arm-embedded.buildPackages
false

nix-repl> pkgsCross.aarch64-multiplatform.stdenv.hostPlatform.emulatorAvailable pkgsCross.aarch64-multiplatform.buildPackages
true
```

will be useful for stuff like handling https://github.com/NixOS/nixpkgs/issues/187109

Atemu commented on lib/systems/default.nix:
```diff
@@ -158,38 +158,47 @@ rec {
+          then "${pkgs.mmixware}/bin/mmix"
+          else null;
+      in {
+        emulator = pkgs:
+          if (final.emulatorAvailable pkgs)
+          then emulators pkgs
+          else throw "Don't know how to run ${final.config} executables.";
```
```suggestion
        emulatorAvailable = pkgs: !lib.isNull (emulators pkgs);

        emulator = pkgs:
          if (final.emulatorAvailable pkgs)
          then emulators pkgs
          else throw "Don't know how to run ${final.config} executables.";
```

(and delete line below, GH won't let me do that in a review suggestion.)

Artturin replied:
done but did `(selectEmulator pkgs) != null` instead because `isNull` is "deprecated" (for 10+ years...)

Atemu commented on lib/systems/default.nix:
```diff
@@ -158,38 +158,47 @@ rec {
         else null;
+    } // (
+      let
+        emulators = pkgs:
```
Perhaps call this `selectEmulator` or something. `emulator` and `emulators` is a bit confusing.

Artturin replied:
done

Atemu reviewed (approved):
A few nits but this is useful.

This PR allows for patches like

```patch
diff --git a/pkgs/development/libraries/gobject-introspection/wrapper.nix b/pkgs/development/libraries/gobject-introspection/wrapper.nix
index 4b3fa1a1981..a2e1abb0376 100644
--- a/pkgs/development/libraries/gobject-introspection/wrapper.nix
+++ b/pkgs/development/libraries/gobject-introspection/wrapper.nix
@@ -24,7 +24,9 @@ in
 
     (
       export bash="${buildPackages.bash}"
+    '' + lib.optionalString (stdenv.targetPlatform.emulatorAvailable buildPackages) ''
       export emulator=${lib.escapeShellArg (stdenv.targetPlatform.emulator buildPackages)}
+    '' + ''
       export buildobjdump="${buildPackages.stdenv.cc.bintools}/bin/objdump"
 
       export targetgir="${lib.getDev (targetPackages.gobject-introspection-unwrapped.override argsForTarget)}"
```

that would fix https://github.com/NixOS/nixpkgs/issues/187109.
